src/yaml/from_yaml.cr

```diff
@@ -22,39 +22,86 @@ end
 
 def Nil.new(pull : YAML::PullParser)
   location = pull.location
-  value = pull.read_scalar
-  if value.empty?
+  value = pull.read_plain_scalar
+  if YAML::NULL_VALUES.includes? value
     nil
   else
-    raise YAML::ParseException.new("Expected nil, not #{value}", *location)
+    raise YAML::ParseException.new("Expected null, not '#{value}'", *location)
   end
 end
 
 def Bool.new(pull : YAML::PullParser)
-  pull.read_scalar == "true"
+  location = pull.location
+  value = pull.read_plain_scalar
+  if YAML::TRUE_VALUES.includes? value
+    true
+  elsif YAML::FALSE_VALUES.includes? value
+    false
+  else
+    raise YAML::ParseException.new("Expected boolean, not '#{value}'", *location)
+  end
 end
 
-{% for type in %w(Int8 Int16 Int32 Int64 UInt8 UInt16 UInt32 UInt64) %}
-  def {{type.id}}.new(pull : YAML::PullParser)
+# TODO: Ideally, it may be beter to use `for type in Int::Primitive.union_types`
+# but it is currently broken due to: https://github.com/crystal-lang/crystal/issues/4301
+{% for bits in [8, 16, 32, 64] %}
+  {% for type in %w(Int UInt) %}
+    def {{type.id}}{{bits}}.new(pull : YAML::PullParser)
+      location = pull.location
+      value = pull.read_plain_scalar
+
+      number = {% if type == "UInt" %}
+        value.to_u{{bits}}?(underscore: false, prefix: true)
+      {% else %}
+        value.to_i{{bits}}?(underscore: false, prefix: true)
+      {% end %}
+      return number if number
+
+      raise YAML::ParseException.new("Invalid {{type.id}}{{bits}} number", *location)
+    end
+  {% end %}
+{% end %}
+
+{% for bits in [32, 64] %}
+  def Float{{bits}}.new(pull : YAML::PullParser)
     location = pull.location
-    begin
-      {{type.id}}.new(pull.read_scalar)
-    rescue ex
-      raise YAML::ParseException.new(ex.message.not_nil!, *location)
+    value = pull.read_plain_scalar
+    if float = value.to_f{{bits}}?
```

Comment:
I'd use `case` here.

Reply:
@Sija cant use case in `from_yaml` but went ahead and used in in `to_yaml`

Reply:
Why not? You can omit the expression if that's what you mean.

Reply:
What benefit does `case` give you here. I cannot call methods on the object to determine each one of my cases. I have to call includes? on other objects. I think `case` makes sense when you can use it's syntactic sugar for calling methods or doing type checks.